### ISSUE TITLE
Jdusudjh

### ISSUE THREAD

**Issue by davalillo-rm** [davalillo-rm/davalillo-rm#187]
import numpy as np
import pandas as pd
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt

def simular_emision_pca_cdbcrp(df_curva, tenor_nuevo, shock_bps, n_pc=3):
    # 1. Preparar la matriz histórica de tasas (quitamos 'Fecha' y limpiamos nulos)
    df_pca = df_curva.drop(columns=['Fecha']).dropna()
    nodos_historicos = list(df_pca.columns)
    
    # 2. Aislar la curva más reciente (la de "hoy" o último día de tu base)
    curva_hoy = df_pca.iloc[-1].copy()
    
    # 3. Interpolar el nuevo nodo (ej. 38 días) si no existe en la matriz estándar
    if tenor_nuevo not in nodos_historicos:
        # Interpolamos históricamente para que el PCA entienda cómo se mueve ese nuevo punto
        df_pca[tenor_nuevo] = df_pca.apply(lambda row: np.interp(tenor_nuevo, nodos_historicos, row.values), axis=1)
        df_pca = df_pca.reindex(sorted(df_pca.columns), axis=1)
        
        # Interpolar en la curva de hoy
        curva_hoy.loc[tenor_nuevo] = np.interp(tenor_nuevo, nodos_historicos, curva_hoy.values)
        curva_hoy = curva_hoy.sort_index()

    nodos_finales = list(df_pca.columns)
    
    # 4. Entrenar el PCA con la matriz histórica
    pca = PCA(n_components=n_pc)
    pca.fit(df_pca)
    eigenvectores = pca.components_ 
    
    # 5. Crear el Vector de Shock (0 en todo, excepto en el nuevo nodo)
    vector_shock = np.zeros(len(nodos_finales))
    idx_nuevo = nodos_finales.index(tenor_nuevo)
    vector_shock[idx_nuevo] = shock_bps / 100.0  # Asumiendo que df_curva está en % (ej. 4.60)
    
    # 6. Transformación: Marginal Z-Score = Shock * Eigenvectores (Traspuestos)
    delta_z_scores = np.dot(vector_shock, eigenvectores.T)
    
    # 7. Impacto Distribuido a lo largo de toda la curva (Efecto Contagio)
    impacto_distribuido = np.dot(delta_z_scores, eigenvectores)
    
    # 8. Curva Repriceada = Curva Hoy + Impacto del PCA
    curva_repriceada = curva_hoy.values + impacto_distribuido
    
    # === GRAFICACIÓN ===
    df_resultados = pd.DataFrame({
        'Nodo': nodos_finales,
        'Curva_Original': curva_hoy.values,
        'Curva_Repriceada': curva_repriceada
    })
    
    plt.figure(figsize=(12, 6))
    plt.plot(df_resultados['Nodo'], df_resultados['Curva_Original'], marker='o', label='Curva Pre-Emisión (Ayer)', color='gray', linestyle='--')
    plt.plot(df_resultados['Nodo'], df_resultados['Curva_Repriceada'], marker='x', label=f'Curva Repriceada (+{shock_bps}bps en nodo {tenor_nuevo}d)', color='#ce2d2d', linewidth=2)
    
    plt.axvline(x=tenor_nuevo, color='black', linestyle=':', label='Nuevo Instrumento (Shock)')
    
    plt.title('Repricing de Curva vía Análisis de Componentes Principales (PCA)', fontsize=14, fontweight='bold', pad=15)
    plt.xlabel('Días al Vencimiento (Tenor)')
    plt.ylabel('Tasa (%)')
    plt.legend()
    plt.grid(alpha=0.4)
    plt.tight_layout()
    plt.show()
    
    return df_resultados
